### ISSUE TITLE
Review and improve async patterns

### ISSUE THREAD

**Issue by peske** [jumptrading/FluentTerminal#237]
I've mentioned this issue months ago, and promised to put it here. It's here finally. As the title implies, I'm bit unsatisfied with way how async pattern is implemented in FT, and I'm sure that it can be improved. This won't result in a shiny feature which user will notice, but it has some benefits:

- Cleaner code
- Slight improvement in performances (possibly)
- Maybe improved stability

I believe that it can be done in 2-3 working days.

I won't start with this without approval from stakeholders.

**Comment by mjs** [comment]
Can you provide some more details about what is wrong with how async is done in FT?

**Comment by peske** [comment]
There are several mistake types.

**1) Tasks within another tasks/threads**

First example:
```
Task.Run(async () => await JumpListHelper.Update(_settingsService.GetShellProfiles()));
```
For those who aren't too experienced with .NET I'll explain what's wrong with this line. There are actually several issues with with it, but the most important is creating a new `Task` (with `Task.Run()` method), which launches another `Task` inside (`JumpListHelper.Update()` method, and only waits for it to finish. It would be like telling Max something like "John, tell Jane to do job X, to notify you when she finishes, and then you notify me". The same result can be accomplished by simply telling directly to Jane: "Jane, do job X and notify me when you're done". In other words, the next line of code would have the same result:
```
JumpListHelper.Update(_settingsService.GetShellProfiles());
```
Not only that the code is much simpler, but some runtime overhead is removed: `Task.Run()` creates a new thread, and `async`/`await` lambda is converted internally by compiler to three _regular_ methods. In short, every call begins with `Task.Run(async ...` or `Task.Factory.StartNew(asyc ...` is simply wrong. I can't come up with a use case where launching a `Task` whose only purpose is to launch another `Task` is useful.

**2) Using `Task.Run()` method**

Secondly, even if the new thread is needed, it should be started with `Task.Factory.StartNew()`, not with `Task.Run()`. These methods have the same behavior, but the difference is that the later is a lower-level abstraction which allows more control by providing many other optional input parameters whose purpose is to configure the new `Task` in the appropriate way. As I can see, we never provide these optional parameters. `Task.Factory.StartNew()`, on the other side, is a higher-level abstraction which basically wraps `Task.Run()` and configures it with additional parameters in a way that is the most optimal for majority of use cases. In other words, `Task.Factory.StartNew()` is the go-to way, while `Task.Run()` is only for advanced usage when we actually know that we should configure some parameters for particular, unusual use case.

We have few dozens of `Task.Run()` usages throughout the code.

**3) Asynchronous programming vs multi-threading**

Many people uses these terms interchangeably, although they are quite different. The best proof for this statement is fact that JavaScript supports asynchronous programming, although it has only one thread available in browsers (the UI thread). 

When we are using `Task.Run()` or `Task.Factory.StartNew()`, we're actually creating a new thread (taking one from the internal thread pool). In some cases this is justifiable, of course. The problem is that most often we're creating a new thread whose only purpose is to wait for another task to finish. This is like telling "John, I have a job for you: wait here and do nothing."

Rule of thumb is to use new threads only for CPU-intensive tasks, when something actually needs to be done (calculated, etc.). Other tasks, which are simply waiting for something (like waiting a data from network, file content from HDD, etc.) should be implemented without threads (so-called _threadless tasks_).

In previous point I've explained that we should replace `Task.Run()` with `Task.Factory.StartNew()`, but I've also noted there _even if the new thread is needed_. In our case the new thread often isn't needed, meaning that often we won't replace `Task.Run()`, but remove it completely.

More issues in the new comment below...

**Comment by peske** [comment]
...continuation

**4) CS4014 Warning**

As you've may noticed, we have about 20 CS4014 warnings while compiling the project in Visual Studio. These warnings are about async method calls which aren't awaited. In some cases it is actually planned (i.e. _fire-and-forget_ pattern), but often they should be awaited.

**5) Use of `ConfigureAwait()` method**

Here's an example of bad code:
```
private async void OnShowSettingsRequested(object sender, EventArgs e)
{
    await ShowSettings().ConfigureAwait(true);
}
```

The purpose of `ConfigureAwait()` call is to define which thread should execute the rest of the method. Let's say that the calling code calls an asynchronous method from some thread `Id1`. The part of the method before awaited line (before `await` keyword) will be executed in the same thread, of course. But when awaited call happens, method returns to the caller, and the caller himself continues to use thread `Id1`. Once awaited call is finished, it's time to execute the rest of the method (after `await` line). Regarding this we have two options: to require that the rest of the method is executed in the same thread as the first part (thread `Id1`), in which case we'll append `.ConfigureAwait(true)` to the end of the awaited call, or we don't care which thread will execute the rest, in which case we'll append `.ConfigureAwait(false)`. Obviously the second option is more relaxing for the system because it can grab any free thread. The first option, on the other hand, puts some overhead because the system has to wait for the same exact thread to be available. Again, in some situations it's justifiable, especially in UI apps, for example, if the method is called from the UI thread, and the second part of the method also touches UI elements. But sometimes it's simply unnecessary burden. In the example above we are setting `.ConfigureAwait(true)` although there isn't second part of the method at all (there's nothing below `await` line). In this particular case we obviously don't have to require the same thread. 

Note that this particular issue isn't too significant because `ConfigureAwait(true)` is the default behavior in .NET meaning that it's used almost everywhere.

**6) Scheduling actions to dispatcher**

As we all know, any code that touches UI elements has to be executed from the UI thread. For this reason, when we want to access UI from a background thread (or an unknown thread), we are using dispatcher. What we basically do is _"Hey dispatcher, here's a piece of work that needs to be done in the UI thread. Please schedule it for execution and return me the result once it's executed."_ What effectively happens is that our piece of work gets put in some queue, and then event loop executes all scheduled pieces one-by-one in UI thread, based on priority.

So where's the problem? The problem is that in one particular case we don't have to use the dispatcher: if we are already in UI thread. In this case, instead of putting a piece of work as additional entity to the queue, we can simply immediately execute this piece of work (we are already in UI thread). So, instead of blindly pushing everything to dispatcher we can do something like _"Let me check if I'm in UI thread, and if this is the case I can immediately do this action. If not - I'll bother dispatcher."_

If this is all, the fix would be very easy, but sadly it isn't all... There's also a case in which we are in the UI thread, but still we don't won't to execute the piece of work immediately, but schedule it to dispatcher anyway. This is the case, for example, if the piece of work isn't too important to us, and we don't have a time to wait it's execution. In such cases we'll push it to dispatcher anyway, basically telling _"Dispatcher, please do this when you're idle. I don't have time to wait now."_
For this exact reason I've changed the method that schedules tasks long time ago, so that now it looks like:
```
public Task RunOnDispatcherThread(Action action, bool enforceNewSchedule = true)
{
    if (!enforceNewSchedule && _dispatcher.HasThreadAccess)
    {
        action();

        return Task.CompletedTask;
    }

    return _dispatcher.RunAsync(CoreDispatcherPriority.Normal, () => action()).AsTask();
}
```

The problem is that I introduced parameter `enforceNewSchedule` when I changed the method, meaning that all previous code doesn't use it, meaning that still actions get always scheduled to dispatcher. What I want to do is to take a look at every call, adjust the param appropriately, and test.

Better use of the dispatcher can result in slightly better responsiveness of the UI.

**Comment by peske** [comment]
Finally, I need to repeat what I've said above in the original issue description: Although slight improvement in UI responsiveness may happen, **I don't expect it to be significant, nor even noticeable.** The purpose of this change is to have cleaner, principally correct code, not actual improvement from end user's perspective.

**Comment by mjs** [comment]
Go ahead with these. They are all worth sorting out. 

Regarding the CS4014 Warning, is there a way to disable the warnings for the cases where a fire-and-forget is intentional?

**Comment by peske** [comment]
Yes, we can disable particular CS4014 warning simply by adding some `#pragma` or comment, assuming that we've intentionally used _fire-and-forget_ pattern.

**Comment by peske** [comment]
I've done everything that I've planned. Closing this one.
